packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -85,123 +109,507 @@ const propTypes = {
+          dayInputRef.current.focus();
+        } else {
+          yearInputRef.current.focus();
+        }
+      }
+    } else if (dateOrder === 'DMY') {
+      if (inputValue.length === 2) {
+        if (type === DateUtil.inputType.DAY) {
+          monthInputRef.current.focus();
+        } else {
+          yearInputRef.current.focus();
+        }
+      }
+    } else if (dateOrder === 'YMD') {
+      if (inputValue.length === 4) {
+        monthInputRef.current.focus();
+      } else if (inputValue.length === 2 && type === DateUtil.inputType.MONTH) {
+        dayInputRef.current.focus();
+      }
+    }
+  };
+
+  /**
+   * Sets state for the day, month, and year.
+   * @param {string} inputValue - The input value to set in state.
+   * @param {number} type - The inputType (day, month, or year).
+   */
+  const setDate = (inputValue, type) => {
+    if (type === DateUtil.inputType.DAY) {
+      dateDispatch({ day: inputValue });
+    } else if (type === DateUtil.inputType.MONTH) {
+      dateDispatch({ month: inputValue });
+    } else {
+      dateDispatch({ year: inputValue });
+    }
+
+    moveFocusOnChange(inputValue, type, dateFormatOrder);
+  };
 
-    this.uuid = uuidv4();
+  const handleDateChange = (event, inputValue, type) => {
+    /**
+     * Sets the day, month and year based on input values, formats them
+     * based on the date format variant, and passes the formatted date to onChange.
+     */
+    const day = type === DateUtil.inputType.DAY ? inputValue : date.day;
+    const month = type === DateUtil.inputType.MONTH ? inputValue : date.month;
+    const year = type === DateUtil.inputType.YEAR ? inputValue : date.year;
+
+    if (type === DateUtil.inputType.DAY) {
+      setDayInitialFocused(false);
+    } else if (type === DateUtil.inputType.MONTH) {
+      setMonthInitialFocused(false);
+    } else {
+      setYearInitialFocused(false);
+    }
 
-    this.handleOnButtonClick = this.handleOnButtonClick.bind(this);
-    this.handleOnChange = this.handleOnChange.bind(this);
-    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);
-  }
+    let inputDate;
+    let formattedDate;
+    if (day.length === 2 && month.length === 2 && year.length === 4) {
+      inputDate = DateUtil.convertToISO8601(`${year}-${month}-${day}`, DateUtil.ISO_EXTENDED_DATE_FORMAT);
+      formattedDate = DateUtil.strictFormatISODate(inputDate, momentDateFormat);
+    }
 
-  componentDidUpdate(prevProps) {
-    if (this.shouldShowPicker && !prevProps.shouldShowPicker && this.props.onClick) {
-      this.props.onClick();
+    if (onChange) {
+      if (DateUtil.isValidDate(formattedDate, momentDateFormat)) {
+        onChange(event, formattedDate);
+      } else {
+        let dateString;
+        if (dateFormatOrder === 'MDY') {
+          dateString = month + day + year;
+        } else if (dateFormatOrder === 'DMY') {
+          dateString = day + month + year;
+        } else {
+          dateString = year + month + day;
+        }
```

Comment:
can we move this logic to dateutils ?

Reply:
This is the only place this logic is used and I presumed that we only have logic that is reused in DateUtils.

Reply:
but this logic looks like it could be reused in future too.